ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+    if span._parent is not None:
+        _tag_openai_token_usage(span._parent, llm_output, propagated_cost=propagated_cost + total_cost, propagate=True)
+
+
+@with_traced_module
+def traced_llm_generate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type
+    prompts = args[0]
+    integration = langchain._datadog_integration
+    span = integration.trace(pin, "%s.%s" % (instance.__module__, instance.__class__.__name__))
+    try:
+        if integration.is_pc_sampled_span(span):
+            for idx, prompt in enumerate(prompts):
+                span.set_tag_str("langchain.request.prompts.%d" % idx, integration.trunc(str(prompt)))
+        model = _extract_model_name(instance)
+        if model is not None:
+            span.set_tag_str("langchain.request.model", model)
+        span.set_tag_str("langchain.request.provider", llm_provider)
+        span.set_tag_str("langchain.request.type", "llm")
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
+            else:
+                span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
+        span.set_tag_str("langchain.request.api_key", str(_extract_api_key(instance)))
+
+        completions = func(*args, **kwargs)
+        if isinstance(instance, langchain.llms.OpenAI):
+            _tag_openai_token_usage(span, completions.llm_output)
+            integration.record_usage(span, completions.llm_output)
+
+        for idx, completion in enumerate(completions.generations):
+            if integration.is_pc_sampled_span(span):
+                span.set_tag_str("langchain.response.completions.%d.text" % idx, integration.trunc(completion[0].text))
+            if completion[0].generation_info is not None:
+                span.set_tag_str(
+                    "langchain.response.completions.%d.finish_reason" % idx,
+                    str(completion[0].generation_info.get("finish_reason")),
+                )
+                span.set_tag_str(
+                    "langchain.response.completions.%d.logprobs" % idx,
+                    str(completion[0].generation_info.get("logprobs")),
+                )
+        if integration.is_pc_sampled_log(span):
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s" % (instance.__module__, instance.__class__.__name__),
+                attrs={
+                    "prompts": prompts,
+                    "choices": [
+                        [{"text": completion.text} for completion in completions]
+                        for completions in completions.generations
+                    ],
+                },
+            )
+    except Exception:
```

Comment:
Should we just use a context manager here:

Ex:

```
with integration.trace(pin, "%s.%s" % (instance.__module__, instance.__class__.__name__)) as span:
  ....... this method .....
```

I guess in this case need a new way to compute error metrics.